apps/linear_algebra/tests/test_halide_blas.cpp

```diff
@@ -0,0 +1,280 @@
+        std::unique_ptr<Vector> ex(randomVector(N));    \
+        std::unique_ptr<Vector> ey(randomVector(N));    \
+        std::unique_ptr<Matrix> eA(randomMatrix(N));    \
+        Vector ax(*ex), ay(*ey);                        \
+        Matrix aA(*eA);                                 \
+                                                        \
+        {                                               \
+            Scalar *x = &(ex->front());                 \
+            Scalar *y = &(ey->front());                 \
+            Scalar *A = &(eA->front());                 \
+            cblas_code;                                 \
+        }                                               \
+                                                        \
+        {                                               \
+            Scalar *x = &(ax.front());                  \
+            Scalar *y = &(ay.front());                  \
+            Scalar *A = &(aA.front());                  \
+            hblas_code;                                 \
+        }                                               \
+                                                        \
+        return compareVectors(N, *ey, ay);              \
+    }
+
+#define L3_TEST(method, cblas_code, hblas_code)         \
+    bool test_##method(int N) {                         \
+        Scalar alpha = randomScalar();                  \
+        Scalar beta = randomScalar();                   \
+        std::unique_ptr<Matrix> eA(randomMatrix(N));    \
+        std::unique_ptr<Matrix> eB(randomMatrix(N));    \
+        std::unique_ptr<Matrix> eC(randomMatrix(N));    \
+        Matrix aA(*eA), aB(*eB), aC(*eC);               \
+                                                        \
+        {                                               \
+            Scalar *A = &(eA->front());                 \
+            Scalar *B = &(eB->front());                 \
+            Scalar *C = &(eC->front());                 \
+            cblas_code;                                 \
+        }                                               \
+                                                        \
+        {                                               \
+            Scalar *A = &(aA.front());                  \
+            Scalar *B = &(aB.front());                  \
+            Scalar *C = &(aC.front());                  \
+            hblas_code;                                 \
+        }                                               \
+                                                        \
+        return compareMatrices(N, *eC, aC);             \
+    }
+
+
+template<class T>
+struct BLASTestBase {
+    typedef T Scalar;
+    typedef std::vector<T> Vector;
+    typedef std::vector<T> Matrix;
+
+    std::random_device rand_dev;
+    std::default_random_engine rand_eng;
+
+    BLASTestBase() : rand_eng(rand_dev()) {}
+
+    Scalar randomScalar() {
```

Comment:
random_scalar, random_vector, random_matrix would be more conformant with the usual Halide style.